### PR TITLE
feat(support): fetch data concurrently

### PR DESCRIPTION
**Pull request by johnstcn**
Fetches data concurrently to speed up process of generating a support bundle.

### CODE REVIEW

File: support/support.go [hunk 1]
```diff
@@ -61,84 +62,141 @@ type Deps struct {
...
-	if err != nil {
-		log.Error(ctx, "fetch build info", slog.Error(err))
-	} else {
+	var (
+		d  Deployment
+		m  sync.Mutex
+		wg sync.WaitGroup
+	)
+
+	wg.Add(1)
```
Comment (mafredri) [inline support/support.go:71]: Using [`errgroup.WithContext`](https://pkg.go.dev/golang.org/x/sync/errgroup#WithContext) would be a nice simplification here and allow early exit on error. Simplifies error handling too.

File: support/support.go [hunk 2]
```diff
@@ -61,84 +62,141 @@ type Deps struct {
...
+	go func() {
+		defer wg.Done()
+		bi, err := client.BuildInfo(ctx)
+		if err != nil {
+			log.Error(ctx, "fetch build info", slog.Error(err))
+			return
+		}
+		m.Lock()
```
Comment (mafredri) [inline support/support.go:79]: We don't need a mutex here as each routine is assigning to their own field.

**Review by mafredri — COMMENTED** [review]
I think it works as-is, but could be simplified quite a bit with errgroup.

File: support/support.go (hunk 2 above)
Comment (johnstcn) [reply support/support.go]: Removed

File: support/support.go (hunk 1 above)
Comment (johnstcn) [reply support/support.go]: Updated!

File: support/support.go [hunk 3]
```diff
@@ -225,9 +265,35 @@ func Run(ctx context.Context, d *Deps) (*Bundle, error) {
 		}
 	}
 
-	b.Deployment = DeploymentInfo(ctx, d.Client, d.Log)
-	b.Workspace = WorkspaceInfo(ctx, d.Client, d.Log, d.WorkspaceID, d.AgentID)
-	b.Network = NetworkInfo(ctx, d.Client, d.Log, d.AgentID)
+	var (
+		wg sync.WaitGroup
+		m  sync.Mutex
+	)
```
Comment (mafredri) [inline support/support.go:271]: Switch for errgroup too?

File: support/support.go [hunk 4]
```diff
@@ -61,34 +63,47 @@ type Deps struct {
...
 
 func DeploymentInfo(ctx context.Context, client *codersdk.Client, log slog.Logger) Deployment {
+	// Note: each goroutine assigns to a different struct field, hence no mutex.
 	var d Deployment
-
-	bi, err := client.BuildInfo(ctx)
-	if err != nil {
-		log.Error(ctx, "fetch build info", slog.Error(err))
-	} else {
+	eg, ctx := errgroup.WithContext(ctx)
```
Comment (mafredri) [inline support/support.go:68]: Just realized we only log errors, so we might be interested in partial results?

If so, might make sense to just use `eg := &errgroup.Group{}`. This way the other Go routines won't be interrupted when one runs into an error. (Sorry I didn't take this into account in my previous review.)

**Review by mafredri — APPROVED** [review]

**Review by mtojek — APPROVED** [review]
👍 

out of curiosity: how much time do we save by switching to the concurrent mode?

**Comment by johnstcn** [comment]
> 👍
> 
> out of curiosity: how much time do we save by switching to the concurrent mode?

We're going from about 13 serial requests right now to potentially having 12/13 of those in parallel. So quite a bit, depending on round trip time.

**Comment by mtojek** [comment]
I hope we don't hit any Coder rate limiter or customer's proxy :)

**Comment by johnstcn** [comment]
> I hope we don't hit any Coder rate limiter or customer's proxy :)

Fair, I'll set max concurrency to 4 as our default rate limit is 8 requests and change per second.

**Review by dannykopping — APPROVED** [review]
Neat 👍 
Do we have any idea how long these steps take in practice?
Do we need timing info for each step? Presumably we're parallelising the steps because the fetching currently takes too long.

**Comment by dannykopping** [comment]
> > I hope we don't hit any Coder rate limiter or customer's proxy :)
> 
> Fair, I'll set max concurrency to 4 as our default rate limit is 8 requests and change per second.
> 
> Edit: actually, if folks run into issues with this we can just set GOMAXPROCS

Something to keep in mind:
`GOMAXPROCS` defaults to `# of processors` unless explicitly set, and since this will presumably be run on beefy servers we can expect this figure to be quite high (> than the 13 steps, surely).

We should probably use https://github.com/uber-go/automaxprocs unless it's been disregarded previously.

**Comment by johnstcn** [comment]
> We should probably use https://github.com/uber-go/automaxprocs unless it's been disregarded previously.

I've been on the fence about adding this; it's been bugging me for a while that the Go runtime isn't CFS-aware. It would be a big enough change though. Maybe a separate PR?

**Comment by mafredri** [comment]
We had a [PR for enabling automaxprocs in the agent](https://github.com/coder/coder/pull/10594), but it seems to have fallen into oblivion. I'm all for it, though.

**Pull request merged**